### PR TITLE
[tune] Let Search Algorithms use early stopped trials 

### LINKED ISSUE #5459: [tune] Let Search Algorithms use early stopped trials
### Describe the problem
All search algorithms provided should record the last seen score of the trial, even when the trial is early stopped. This should boost performance and allow schedulers/search algorithms to interact.

### PR DESCRIPTION
**Pull request by adizim**
Enables Search Algorithms to use early stopped trials (early stopped by trial schedulers). This can greatly accelerate the optimization process.

## Related issue number

 - [x] Fix sigopt + hyperopt early stopping
 - [x] Add proper docstrings
 - [x] Compare `use_early_terminated_trials` with `hyperopt_example.py`. 
 
Closes #5459

## TODOs after this PR:

Add tests for all search algorithms.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.

### CODE REVIEW

**Comment by richardliaw** [comment]
Thanks for making this PR! 

So the correct way to do this is to pass in `result` when the trial is stopped by the trial scheduler. This should be just done in the trial scheduler. You may also want to pass a flag in to search algorithms in case users don't want to use these early terminated results.

File: python/ray/tune/suggest/skopt.py [hunk 1]
```diff
@@ -151,9 +152,14 @@ def on_trial_complete(self,
         as it minimizes on default.
         """
         skopt_trial_info = self._live_trial_mapping.pop(trial_id)
+        if early_terminated:
```
Comment (ericl) [inline python/ray/tune/suggest/skopt.py]:
```suggestion
        if early_terminated and self.use_early_stopped:
```

**Review by ericl — COMMENTED** [review]
Looks good, but should it be an option?

**Comment by adizim** [comment]
A few questions for this task:

1. for the trial scheduler, we need to change it so that the result gets used even on trial stop. To do that I have edited the `on_trial_result` method, b/c from what I understand the `on_trial_complete` method will **not** trigger on trial stops. Does that sound right?
2. For search algorithms, is `on_trial_complete` triggered on trial stops? I assume it is, because we pass in the `early_terminated` flag for `on_trial_complete`

File: python/ray/tune/trial_runner.py [hunk 2]
```diff
@@ -674,9 +676,18 @@ def stop_trial(self, trial):
...
         elif trial.status in [Trial.PENDING, Trial.PAUSED]:
-            self._scheduler_alg.on_trial_remove(self, trial)
-            self._search_alg.on_trial_complete(
-                trial.trial_id, early_terminated=True)
+            try:
+                result = self.trial_executor.fetch_result(trial)
+                trial.update_last_result(result, terminate=True)
+                self._scheduler_alg.on_trial_remove(self, trial)
+                self._search_alg.on_trial_complete(
+                    trial.trial_id, result=result, early_terminated=True)
```
Comment (richardliaw) [inline python/ray/tune/trial_runner.py]: I think we should probably leave this as is; calling `stop_trial` should probably not block.

File: python/ray/tune/suggest/bayesopt.py [hunk 3]
```diff
@@ -102,6 +106,8 @@ def on_trial_complete(self,
                           error=False,
                           early_terminated=False):
         """Passes the result to BayesOpt unless early terminated or errored"""
+        if early_terminated and self.use_early_terminated_trials is False:
+            return
         if result:
```
Comment (richardliaw) [inline python/ray/tune/suggest/bayesopt.py]: this should probably be like; 

```
if result:
	if early_terminated and self.use_early_terminated_trials is False:
        return
    ...
```

Can you also do this for all of the search algorithms?

Thanks!

File: python/ray/tune/suggest/bayesopt.py [hunk 4]
```diff
@@ -75,6 +78,7 @@ def __init__(self,
         elif mode == "min":
             self._metric_op = -1.
         self._live_trial_mapping = {}
+        self._use_early_terminated_trials = use_early_terminated_trials
```
Comment (richardliaw) [inline python/ray/tune/suggest/bayesopt.py]: can you set this at `SuggestionAlgorithm`?

File: python/ray/tune/suggest/hyperopt.py [hunk 5]
```diff
@@ -183,9 +185,6 @@ def on_trial_complete(self,
         if error:
             ho_trial["state"] = hpo.base.JOB_STATE_ERROR
             ho_trial["misc"]["error"] = (str(TuneError), "Tune Error")
-        elif early_terminated:
```
Comment (richardliaw) [inline python/ray/tune/suggest/hyperopt.py]: @adizim `use_early_terminated_trials` should be used here too right?

**Review by richardliaw — COMMENTED** [review]
Can you raise a warning in `sigopt.py` if `use_early_terminated_trials` is in the `kwargs` that the parameter is not used in the class?

Also, the last thing to do is to compare `hyperopt_example.py` on before and after.

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by shaneknapp** [comment]
jenkins test tune

**Comment by shaneknapp** [comment]
ok to test

**Comment by shaneknapp** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
jenkins test tune

**Review by richardliaw — APPROVED** [review]

**Pull request merged**